src/hypothesis/strategies.py

```diff
@@ -857,18 +857,21 @@ def recursive(base, extend, max_leaves=100):
 def permutations(values):
     """Return a strategy which returns permutations of the collection
     "values"."""
+    from hypothesis.internal.conjecture.utils import integer_range
+
     values = list(values)
     if not values:
-        return just(()).map(lambda _: [])
-
-    def build_permutation(swaps):
-        initial = list(values)
-        for i, j in swaps:
-            initial[i], initial[j] = initial[j], initial[i]
-        return initial
-    n = len(values)
-    index = integers(0, n - 1)
-    return lists(tuples(index, index), max_size=n ** 2).map(build_permutation)
+        return builds(list)
+
+    class PermutationStrategy(SearchStrategy):
+
+        def do_draw(self, data):
+            result = list(values)
+            for i in range(len(result)):
+                j = integer_range(data, i, len(result) - 1)
```

Comment:
If I’m reading this correctly (and I could have misread `integer_range()`), doesn’t this return an integer in `[i + 1, len(result) - 1]`? So subsequent iterations will only shuffle the latter elements of the list, whereas the original could swap any two elements of the list at any point in the iteration.

I’m not convinced this makes a difference, but I’m also not convinced it _doesn’t_ make a difference.

Thoughts?